internal/app/backend/backend_service.go

```diff
@@ -145,53 +172,32 @@ func (s *backendService) getGRPCClient(config *pb.FunctionConfig_Grpc) (pb.Match
 // matchesFromGrpcMMF triggers execution of MMFs to fetch match results for each profile.
 // These proposals are then sent to evaluator and the results are streamed back on the channel
 // that this function returns to the caller.
-func (s *backendService) matchesFromGrpcMMF(ctx context.Context, profiles []*pb.MatchProfile, client pb.MatchFunctionClient) <-chan *pb.Match {
-	c := make(chan *pb.Match)
-
-	// Create a goroutine that will wait for completion of individual goroutines processing each
-	// match profile and then close the channel signaling completion of match results generation.
-	go func(c chan<- *pb.Match) {
-		var wg sync.WaitGroup
-		for _, profile := range profiles {
-			// Trigger a goroutine for each profile. This will call the MatchFunction to get
-			// proposed matches and will then pass these to evaluator to get the accepted
-			// match results that it will send on the output channel.
-			wg.Add(1)
-			go func(profile *pb.MatchProfile) {
-				defer wg.Done()
-				// TODO: This code calls user code and could hang. We need to add a deadline here
-				// and timeout gracefully to ensure that the ListMatches completes.
-				// TODO: Currently, a failure in running the MMF is silently ignored and does not
-				// fail the FetchMatches. This needs to be revisited to investigate whether the
-				// FetchMatches should fail or atleast hint that a part of MMF executions failed.
-				resp, err := client.Run(ctx, &pb.RunRequest{Profile: profile})
-				if err != nil {
-					logger.WithFields(logrus.Fields{
-						"error":   err.Error(),
-						"profile": profile,
-					}).Error("failed to run match function for profile")
-					return
-				}
-
-				logger.WithFields(logrus.Fields{
-					"profile":   profile,
-					"proposals": resp.Proposal,
-				}).Trace("proposals generated for match profile")
-				// TODO: The matches returned by the MatchFunction will be sent to the
-				// Evaluator to select results. Until the evaluator is implemented,
-				// we channel all matches as accepted results.
-				for _, match := range resp.Proposal {
-					c <- match
-				}
-			}(profile)
-		}
-
-		// Wait for all the goroutines handling each profile to return and then close the channel.
-		wg.Wait()
-		close(c)
-	}(c)
+func (s *backendService) matchesFromGrpcMMF(ctx context.Context, profile *pb.MatchProfile, client pb.MatchFunctionClient, matchChan chan<- *pb.Match, errChan chan<- error) {
```

Comment:
nit: Seems like this method is standalone and doesn't need to be a method of this class.